File: 0112-path-sum/0112-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int targetSum) {
        if(!root) return false;
        return temp(root, 0, targetSum);
    }

    bool temp(TreeNode* root, int sum, int targetSum){
        if(!root) return false;
        sum += root->val;
        if(root && sum == targetSum && !root->left && !root->right) return true;

        return temp(root->left, sum, targetSum) ||
            temp(root->right, sum, targetSum);
    }
};

File: 0112-path-sum/0112-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean hasPathSum(TreeNode root, int targetSum) {
        if(root == null) return false;
        return hasPath(root, targetSum);
    }

    private boolean hasPath(TreeNode node, int sum){
        if(node.left == null && node.right == null) return sum - node.val == 0;
        else if(node.left == null) {
            return hasPath(node.right, sum - node.val);
        } else if(node.right == null){
            return hasPath(node.left, sum - node.val);
        } else {
            return hasPath(node.left, sum - node.val) || 
                hasPath(node.right, sum - node.val) ;
        }
    }
}